src/coap_tinydtls.c

```diff
@@ -201,30 +192,109 @@ get_psk_info(struct dtls_context_t *dtls_context,
   switch (type) {
   case DTLS_PSK_IDENTITY:
 
-    if (id_len)
-      coap_log(LOG_DEBUG, "got psk_identity_hint: '%.*s'\n", (int)id_len, id);
-
-    if (!coap_context || !coap_context->get_client_psk)
+    if (!coap_context || !coap_context->get_client_psk ||
+        coap_session->type != COAP_SESSION_TYPE_CLIENT)
       goto error;
 
+    setup_cdata = &coap_session->cpsk_setup_data;
+
+    temp.s = id;
+    temp.length = id_len;
+    coap_session_refresh_psk_hint(coap_session, &temp);
+
+    coap_log(LOG_DEBUG, "got psk_identity_hint: '%.*s'\n", (int)id_len, id ? (const char*)id : "");
+
+    if (setup_cdata->validate_ih_call_back) {
+      coap_str_const_t lhint;
+      lhint.length = id_len;
+      lhint.s = id;
+      const coap_dtls_cpsk_key_t *psk_key =
+               setup_cdata->validate_ih_call_back(&lhint,
+                                                  coap_session,
+                                                  setup_cdata->ih_call_back_arg);
+
+      if (psk_key == NULL)
+        return 0;
+      if (psk_key->identity.length >= result_length)
+        return 0;
+      if (psk_key->key.length > sizeof(psk))
+        return 0;
+
+      if (coap_session->psk_identity) {
+        coap_delete_bin_const(coap_session->psk_identity);
+      }
+      identity_length = psk_key->identity.length;
+      coap_session->psk_identity = coap_new_bin_const(psk_key->identity.s, identity_length);
+      memcpy(result, psk_key->identity.s, identity_length);
+      result[identity_length] = '\000';
+
+      temp.s = psk_key->key.s;
+      temp.length = psk_key->key.length;
+      coap_session_refresh_psk_key(coap_session, &temp);
+
+      return identity_length;
+    }
+
     identity_length = 0;
     psk_len = coap_context->get_client_psk(coap_session, (const uint8_t*)id, id_len, (uint8_t*)result, &identity_length, result_length, psk, sizeof(psk));
     if (!psk_len) {
       coap_log(LOG_WARNING, "no PSK identity for given realm\n");
       fatal_error = DTLS_ALERT_CLOSE_NOTIFY;
       goto error;
     }
-    client = 1;
     return (int)identity_length;
 
   case DTLS_PSK_KEY:
-    if (coap_context->get_server_psk)
+    if (coap_session->type == COAP_SESSION_TYPE_CLIENT) {
+      if (!coap_context || !coap_context->get_client_psk)
+        goto error;
+      identity_length = 0;
+      psk_len = coap_context->get_client_psk(coap_session, (const uint8_t*)id, id_len, (uint8_t*)result, &identity_length, result_length, psk, sizeof(psk));
+      if (!psk_len) {
+        coap_log(LOG_WARNING, "no pre-shared key for given realm\n");
+        fatal_error = DTLS_ALERT_CLOSE_NOTIFY;
+        goto error;
+      }
+      memcpy(result, psk, psk_len);
```

Comment:
I fail to see where `psk_len <= result_length` is ensured.

Reply:
Good catch - check is now back in.